aptos-move/framework/aptos-framework/sources/stake.move

```diff
@@ -1061,194 +1062,152 @@ module aptos_framework::stake {
-                successful_proposals: 0,
-                failed_proposals: 0,
-            });
+        if (commit) {
+            *validator_set = new_validator_set;
 
-            // Automatically renew a validator's lockup for validators that will still be in the validator set in the
-            // next epoch.
-            let stake_pool = borrow_global_mut<StakePool>(validator_info.addr);
-            if (stake_pool.locked_until_secs <= timestamp::now_seconds()) {
+            // Also update validator indices, reset performance scores, and renew lockups.
+            validator_perf.validators = vector::empty();
+            let recurring_lockup_duration_secs = staking_config::get_recurring_lockup_duration(&config);
+            let vlen = vector::length(&validator_set.active_validators);
+            let validator_index = 0;
+            while ({
                 spec {
-                    assume timestamp::spec_now_seconds() + recurring_lockup_duration_secs <= MAX_U64;
+                    invariant spec_validators_are_initialized(validator_set.active_validators);
+                    invariant len(validator_set.pending_active) == 0;
+                    invariant len(validator_set.pending_inactive) == 0;
+                    invariant 0 <= validator_index && validator_index <= vlen;
+                    invariant vlen == len(validator_set.active_validators);
+                    invariant forall i in 0..validator_index:
+                        global<ValidatorConfig>(validator_set.active_validators[i].addr).validator_index < validator_index;
+                    invariant len(validator_perf.validators) == validator_index;
                 };
-                stake_pool.locked_until_secs =
-                    timestamp::now_seconds() + recurring_lockup_duration_secs;
+                validator_index < vlen
+            }) {
+                let validator_info = vector::borrow_mut(&mut validator_set.active_validators, validator_index);
+                let validator_config = borrow_global_mut<ValidatorConfig>(validator_info.addr);
+                validator_config.validator_index = validator_index;
+
+                vector::push_back(&mut validator_perf.validators, IndividualValidatorPerformance {
+                    successful_proposals: 0,
+                    failed_proposals: 0,
+                });
+
+                // Automatically renew a validator's lockup for validators that will still be in the validator set in the
+                // next epoch.
+                let stake_pool = borrow_global_mut<StakePool>(validator_info.addr);
+                if (stake_pool.locked_until_secs <= reconfiguration_state::start_time_secs()) {
+                    spec {
+                        assume reconfiguration_state::start_time_secs() + recurring_lockup_duration_secs <= MAX_U64;
+                    };
+                    stake_pool.locked_until_secs =
```

Comment:
Can you elaborate on why this is required and what the impact/implications (if any) would be?

Reply:
imagine:
12:34:00 DKG start with validator Alice(active_stake=30, pending_inactive=50, locked_until=12:34:15) // randomness_power=80
12:34:20 DKG finish, where we update_stake_pool() for Alice.

in `update_stake_pool()` if we look at "current time", Alice with enter new epoch with actual stake=30 and randomness power=80, which is incorrect.

Reply:
Unlike fee/rewards which is considered negligible, I assume stake changes like this are much more larger, and we have to deal with this to make sure randomness is secure.